content/post/smaller-the-parameter-smaller-the-bugs.md

```diff
@@ -0,0 +1,242 @@
+#### Old pattern without refactoring
+```rb
+def summary(first_name, last_name, middle_name = nil)
+  if middle_name.nil?
+    "#{title} - #{last_name}, #{first_name} (#{publish_year})"
+  else
+    "#{title} - #{last_name}, #{middle_name}, #{first_name} (#{publish_year})"
+  end
+end
+```
+#### After refactoring
+```rb
+def summary
+  "#{title} - #{author.full_name} (#{publish_year})"
+end
+```
+Obviously, the code quality has a huge improvement after refactoring. 
+## Loose coupling
+Before we finish this refactoring and send the PR to someone, let's check the code again, does it clean enough?
+
+The code seems short and simple, but this method still depends on the author class. In other words, the `summary` method need to know the author object has a `full_name` method, we could say the book class still couple with author class. In this case, maybe it's hard to notice the cost of coupling, let's imagine after few years later, the `summary` method might become more complicated:
+
+#### Few years later
+```rb
+def summary
+  # 10 lines of code you don't want to read.
+  middle_result = "#{some_author_title_helper(author)} - #{author.full_name}"
+  # another 10 lines of code
+  return final_result
+end
+```
+The disadvantage of coupling is that when one class needs to change, you need to change others as well. For example, one day we need to rename the full name method or adding more parameters to it, we need to modify the book class at the same time, which we don't like.
+
+The bad news is, in the real world it's almost impossible to remove all the coupling between each object, but we can at least reduce the impact by segregate and highlight the dependence. 
+
+```rb
+def summary
+  "#{title} - #{author_name} (#{publish_year})"
+end
+
+def author_name
+  author.full_name
+end
+```
+After this refactoring, the `summary` method 100% calling from the book object itself. The book class still has the coupling with the author class, but it's obvious and easy to maintain. If we only consider the example case, you might wonder is that a little bit over design? Well, I would say that a good pattern is never too early. Now we can push the PR confidently and enjoy the holiday.
+
+## Conclusion
+The good code should be small and simple, of course, the parameter is not the only place we can do our effort to reduce the size. During the refactoring, don't forget to confirm the result by unit test to make sure the final behavior would be the same. People around you may think that it would be time-wasting work because the final behavior remains the same even though developers spent a lot of time on it. But there's the value of code quality. I believe it would perform in development speed and product stability in the long term.
```

Comment:
```suggestion
Good code should be small and simple. Of course, the parameter is not the only place we can place our effort in to reducing its size. During refactoring, don't forget to confirm the result by unit testing to make sure the final behavior is consistent. People around you may think that it would be time-wasting work because the final behavior remains the same even though developers spent a lot of time on it. But there's the value of code quality. I believe this line of thinking would help you perform well in development speed and product stability in the long term.
```